web/src/context/auth-provider.tsx

```diff
@@ -0,0 +1,155 @@
+    message: string;
+  }>;
+  logout: () => void;
+  user: User | null;
+}
+
+const AuthContext = createContext<AuthContext | null>(null);
+
+const key = 'user';
+
+/**
+ * Stores or removes the user object in/from localStorage.
+ * @param {User | null} user - The user object to store or `null` to remove the user.
+ */
+function setStoredUser(user: User | null) {
+  if (user) {
+    localStorage.setItem(key, JSON.stringify(user));
+  } else {
+    localStorage.removeItem(key);
+  }
+}
+
+/**
+ * Provides authentication context to its children.
+ * @param {{children: React.ReactNode}} props - The children components that will receive the auth context.
+ * @returns {JSX.Element} The AuthProvider component.
+ */
+export function AuthProvider({children}: {children: React.ReactNode}) {
+  const [user, setUser] = useState<User | null>(
+    JSON.parse(localStorage.getItem(key) || 'null')
+  );
+
+  const isAuthenticated = !!user;
+
+  /**
+   * Logs out the user by removing the stored user object and setting the user to null.
+   */
+  const logout = () => {
+    setStoredUser(null);
+    setUser(null);
+  };
+
+  /**
+   * Logs in the user with a token and refresh token.
+   * @param {string} token - The token to use for authentication.
+   * @param {string} refreshToken - The refresh token to use for authentication.
+   * @returns {Promise<{status: string, message: string}>} A promise that resolves to an object containing the status and message.
+   */
+  const loginWithToken = async (token?: string, refreshToken = '') => {
+    if (!token) return {status: 'error', message: 'No token provided'};
+
+    const response = await fetch(
+      `${import.meta.env.VITE_API_URL}/api/users/current`,
+      {
+        method: 'GET',
+        headers: {
+          'Content-Type': 'application/json',
+          Authorization: `Bearer ${token}`,
+        },
+      }
+    );
+
+    if (!response.ok) return {status: 'error', message: 'Error fetching user'};
+
+    const user = await response.json();
+
+    setStoredUser({user, token, refreshToken});
+    setUser({user, token, refreshToken});
+
+    return {status: 'success', message: ''};
+  };
+
+  /**
+   * Refreshes the user's token by making a POST request to the API.
+   * @returns {Promise<{status: string, message: string}>} A promise that resolves to an object containing the status and message.
+   */
+  const refreshToken = async () => {
+    if (!user) return {status: 'error', message: 'No user to refresh'};
+
+    const response = await fetch(
+      `${import.meta.env.VITE_API_URL}/api/auth/refresh`,
+      {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+          Authorization: `Bearer ${user.token}`,
+        },
+        body: JSON.stringify({refreshToken: user.refreshToken}),
+      }
+    );
+
+    if (!response.ok) {
+      setStoredUser(null);
+      setUser(null);
+
+      return {status: 'error', message: 'Refresh token failed'};
+    }
```

Comment:
Some issues here I think 

a) only try and refresh the token if the current access token is expiring soon (e.g. <1 minute) (saves network traffic)
b) if the refresh fails, don't immediately boot the user out, instead prompt a refresh of the isAuthenticated state based on the existing access token - if it's not expired yet (or within some margin) then let the user continue working and try the refresh again on the next interval